### PR TITLE
refactored and rewrote most of the plugin

### PR DESCRIPTION
**Pull request by nhooyr**
Mostly everything has been simplified. I removed the `Nman 3 printf`
syntax which I'm gonna have to add back. A bit more testing and I think
we're good to go.

### CODE REVIEW

**Comment by nhooyr** [comment]
I added back the `Nman 3 printf` syntax and massive improvements to the autocompletion (it won't show the section unless the section is not given) and generally the code is much more clean.

Just need to update the docs and we're done :P

**Comment by nhooyr** [comment]
@zchee @Corax26 @lervag 
If you guys find the time, please test it! If this works well, it may become the default man page plugin in neovim!

**Comment by zchee** [comment]
@nhooyr Of course :)

**Comment by nhooyr** [comment]
I included improvements from neovim/neovim#4168, I'm using a global `K` mapping instead of `keywordprg` and documentation has been updated.

I just need to comment the code, a bit more testing, and I think we're good to go!

**Comment by nhooyr** [comment]
Now the Japanese characters look perfect.
screen shot 2016-03-13 at 6 32 53 am

**Comment by zchee** [comment]
@nhooyr Yes!! Thanks for lot of support :)

File: ftplugin/neoman.vim [hunk 1]
```diff
@@ -3,31 +3,24 @@ if exists('b:did_ftplugin')
...
+setl nomodified
+setl readonly
+setl nomodifiable
+setl noexpandtab
+setl tabstop=8
+setl softtabstop=8
+setl shiftwidth=8
+setl nolist
+setl foldcolumn=0
+setl colorcolumn=0
```
Comment (justinmk) [inline ftplugin/neoman.vim]: Abbreviated names in scripts are generally advised against. Abbreviation feature of Vim is for human input (i.e. at the `:` prompt).

Comment (nhooyr) [inline ftplugin/neoman.vim]: fixed.

**Comment by lervag** [comment]
I think this looks good, but one thing: I really don't like that you map `K` globally by default. This should instead be optional (or even manual; it is trivial). I personally map `K` to different things depending on context, and so this global `K` mapping can (probably will) interfere.

I propose to either:
1. Make it optional (invert the `g:neoman_no_k` option), or
2. Remove it alltogether and simply show in the readme (and docs, when they are ready) how to map it locally and globally for users.

**Comment by nhooyr** [comment]
@lervag I agree, I'll remove it and move it into the docs.

**Comment by lervag** [comment]
Great! I didn't read the code in detail, but it looks pretty good to me. I agree with @justinmk about being verbose and using full length command and option names. Though I see you've already updated accordingly. Great work!

**Comment by nhooyr** [comment]
@zchee I fixed the bug with neoman always opening at the bottom as well.

**Comment by zchee** [comment]
@nhooyr Thanks for added workaround :)

I was investigated after that conversation. I found again strange behavior.
In Japanese language man page, it have many empty spaces below shown man page character more than English.

Maybe caused by it. but I have been using comfortable with neoman, I do not care :)

**Comment by nhooyr** [comment]
@zchee can you take a screenshot? I'm almost done the syntax highlighting btw, expect a commit in a few.

**Comment by zchee** [comment]
@nhooyr Ok, but now my PC has the cron-building Chromium. very rampage 😱
I'll update later. Thanks!

**Comment by nhooyr** [comment]
Look good?
screen shot 2016-03-13 at 7 05 02 pm

**Comment by zchee** [comment]
Yes, If always located there, no problem!
Because away from the desk, I will also test later.

**Comment by nhooyr** [comment]
I'm not sure though about supporting multiple languages, you know how the C man pages have a `Synopsis` section? It's not possible to detect `synopsis` in every single language, and there is no other way to detect it. Perhaps a variable that the user sets for their language? Idk, maybe we should support english only.

**Comment by zchee** [comment]
Yes, It is not necessarily the `書式`, Its rules are [JMProject](http://linuxjm.osdn.jp/) (This is the deployment location of the typical man page in Japan).
Maybe not matched FreeBSD Japan man page.

I also think the same multi-language support is likely to be unreasonable.
I'm happy if you implements configurable interface. But if this feature too difficult or reduce the performance, It does not need to be implemented.

Also I'm good at support English only :)

**Comment by nhooyr** [comment]
@zchee I'm gonna leave it the way it is for now, it works fine for english and other languages. I'm gonna add a general interface, a regex you define to detect the synopsis.

Btw which looks better?

screen shot 2016-03-14 at 12 31 57 am
screen shot 2016-03-14 at 12 31 52 am

The one with the white bottom, or the one without the white bottom (on the right at the very bottom, notice the text).

**Comment by zchee** [comment]
@nhooyr Sorry for a late reply.

I think it is displayed by a user, the user should already know the name of the command.
`Linux` `26 Dec 1996` `CREATE MODULE(2)`
Important is the content of man page, If this only one thing color is different, it gives the useless attention to the user only a little.

So all of the colors are the same looks good :)

**Comment by nhooyr** [comment]
@zchee yes I went with the one without the useless color.

In that case, just try testing with as many different manpages you can, I'll create a PR for neovim.

**Comment by zchee** [comment]
@nhooyr Nice :)
Yes, Neovim built-in `Man` plugin does not respect of the original `man` command highlight.

**Comment by lervag** [comment]
@nhooyr Have you had any contact with the Vim community about replacing the Vim plugin for Man?

**Comment by nhooyr** [comment]
@lervag I've had much better luck with contributing to neovim. I'll do that right after.

**Comment by nhooyr** [comment]
@zchee I removed the synopsis setting. I'm using a little bit of vim script to figure out the heading now, seems to work well. No need for the user to configure anything!

**Comment by zchee** [comment]
@nhooyr Awesome! I'll test latest rewrite branch :)

**Comment by nhooyr** [comment]
I'm done otherwise, I'll merge it in later today after testing and further feedback from my neovim PR.

**Comment by nhooyr** [comment]
@lervag I added a local `K` mapping. Is that cool?

**Comment by nhooyr** [comment]
I am gonna cry... For some reason FreeBSD has a extra section between the `NAME` and `SYNOPSIS` which means I can't just always select the second section for the c highlighting...

https://www.freebsd.org/cgi/man.cgi?query=printf&apropos=0&sektion=3&manpath=FreeBSD+10.2-RELEASE&arch=default&format=html

I've added a special case, but I think maybe its a better idea to revert and use `g:neoman_synopsis` again. @zchee thoughts?

For now I'm gonna merge, it looks good.

**Comment by lervag** [comment]
Yes, I think the local mapping is OK and fine.

**Comment by zchee** [comment]
@nhooyr Congrats :tada:

I'll check your suggested FreeBSD special case.

**Comment by nhooyr** [comment]
@zchee thanks. but don't worry about it, I'm gonna add back the user interface variable, `g:neoman_synopsis`. Too many special cases with this approach, I think its better to just let the user fill the regex in for their language(s).

**Comment by ghost** [comment]
@nhooyr Just pulled master, sorry I didn't look at the code but just doing some basic checks something has been broken with the completion (yes still annoying you with that :D): there is no match anymore with `:Nman 1 <Tab>`. Probably the fix is trivial? Thanks for your work anyway!

**Comment by nhooyr** [comment]
@Corax26 hmm not sure what could be wrong :(. Could you please take a quick look at the code? It's commented now and easier to read.

**Comment by nhooyr** [comment]
Never mind found the bug and fixed.

**Comment by ghost** [comment]
Yep fix confirmed, thanks!

**Comment by nhooyr** [comment]
neovim now contains a much better version of this plugin!

neovim/neovim#4449 has been merged!

**Comment by zchee** [comment]
@nhooyr :tada:

**Pull request merged**